features/26-mfa/api.md

```diff
@@ -0,0 +1,882 @@
+
+## GET /mfa/authenticators
+
+### Request
+
+No input is required.
+
+### Response body
+
+```JSON
+{
+  "authenticators": [
+    {
+      "id": "1",
+      "type": "totp",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "display_name": "totp-2edfdca2-3f05-41d2-bfd8-502725ee09a6"
+    },
+    {
+      "id": "2",
+      "type": "oob",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "channel": "sms",
+      "masked_phone": "+85223******"
+    },
+    {
+      "id": "3",
+      "type": "oob",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "channel": "email",
+      "masked_phone": "ab******@example.com"
+    }
+  ]
+}
+```
+
+## DELETE /mfa/authenticator
+
+### Request body
+
+```JSON
+{
+  "authenticator_id": "XXX"
+}
+```
+
+### Response body
+
+No response body.
+
+## POST /mfa/recovery_code/regenerate
+
+### Request body
+
+No request body
+
+### Response body
+
+```JSON
+{
+  "recovery_codes": ["xxx", "yyy"]
+}
+```
+
+## GET /mfa/recovery_code
+
+### Request body
+
+No request body
+
+### Response body
+
+```JSON
+{
+  "recovery_codes": ["xxx", "yyy"]
+}
+```
+
+## POST /mfa/bearer_token/revoke_all
+
+### Request body
+
+No request body
+
+### Response body
+
+No response body
+
+## MFA required error
+
+```JSON
+{
+  "name": "MFARequired",
+  "code": "9999",
+  "message": "MFA required",
+  "info": {
+    "mfa_token": "xxx"
+  }
+}
+```
+
+# SDK API
+
+The following functions are all in the namespace `skygear.auth.mfa`.
+
+```typescript
+// Register TOTP
+
+interface CreateNewTOTPResult {
+  authenticatorID: string;
+  authenticatorType: "totp";
+  secret: string;
+}
+
+interface ActivateTOTPResult {
+  recoveryCodes?: string[];
+}
+
+interface GenerateOTPAuthURIOptions {
+  secret: string;
+  issuer: string;
+  accountName: string;
+}
+
+async function createNewTOTP(displayName?: string): Promise<CreateNewTOTPResult>;
+async function activateTOTP(authenticatorID: string, otp: string): Promise<ActivateTOTPResult>;
+function generateOTPAuthURI(options: GenerateOTPAuthURIOptions): string;
+function generateOTPAuthURIQRCodeImageURL(otpauthURI: string): string;
+
+// Register OOB
+
+type CreateNewOOBOptions = CreateNewOOBSMSOptions | CreateNewOOBEmailOptions;
+
+interface CreateNewOOBSMSOptions {
+  channel: "sms";
+  phone: string;
+}
+
+interface CreateNewOOBEmailOptions {
+  channel: "email";
+  email: string;
+}
+
+interface CreateNewOOBResult {
+  authenticatorID: string;
+  authenticatorType: "oob";
+  channel: "sms" | "email";
+}
+
+interface ActivateOOBResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewOOB(options: CreateNewOOBOptions): Promise<CreateNewOOBResult>;
+async function triggerOOB(authenticatorID?: string): Promise<void>;
+async function activateOOB(authenticatorID: string, code: string): Promise<ActivateOOBResult>;
+
+// Error inspection
+
+function isMFARequiredError(err: unknown): boolean;
+
+// Authenticate
+
+interface TOTPAuthenticateOptions {
+  authenticatorID?: string;
```

Comment:
Per our discussion, in most of the case user will only have 1 authenticator. Maybe the authenticatorID can be optional if there is only one authenticator for the user?